### PR TITLE
Fixed Wayland surface creation

### PR DESCRIPTION
**Pull request by asafran**
Added handle of wayland platform, to run which, instead of XCB extension, you need to specify VK_KHR_WAYLAND_SURFACE_EXTENSION.

From Qt documentation:
_Note: It is up to the component creating the external instance to ensure the necessary extensions are enabled on it. These are: VK_KHR_surface, the WSI-specific VK_KHR*surface that is appropriate for the platform in question, and VK_EXT_debug_report in case QVulkanInstance's debug output redirection is desired._

#21 

Tested on Fedora 37, Qt 5.15.8, Qt 6.2.2, Qt 6.4.2, Wayland and X11 session

### CODE REVIEW

File: src/vsgQt/ViewerWindow.cpp [hunk 1]
```diff
@@ -62,9 +66,17 @@ void ViewerWindow::cleanup()
     if (windowAdapter)
     {
 #if QT_HAS_VULKAN_SUPPORT
-        if (surfaceType() == QSurface::VulkanSurface)
+        auto platform = qGuiApp->platformName();
```
Comment (robertosfield) [inline src/vsgQt/ViewerWindow.cpp]: I am confused why you have added initialize functions to a cleanup() function.

File: CMakeLists.txt [hunk 2]
```diff
@@ -66,6 +66,7 @@ elseif(APPLE)
     add_definitions(-DVK_USE_PLATFORM_MACOS_MVK)
 elseif(UNIX)
     add_definitions(-DVK_USE_PLATFORM_XCB_KHR)
+    add_definitions(-DVK_USE_PLATFORM_WAYLAND_KHR)
```
Comment (robertosfield) [inline CMakeLists.txt:69]: Should we not be testing whether Wayland and Xcb are available before adding definitions to suggest they are support?

File: src/vsgQt/ViewerWindow.cpp (hunk 1 above)
Comment (asafran) [reply src/vsgQt/ViewerWindow.cpp]: Oh, I'm sorry I made a mistake when I splitted the PR. There were few changes and I decided to split it manually from a clean repository.

File: CMakeLists.txt (hunk 2 above)
Comment (asafran) [reply CMakeLists.txt:69]: Qt applications can run on different platforms without having to recompile, thanks to QPA.
VK_USE_PLATFORM_WAYLAND_KHR just includes the vulkan header with the wayland surface extension name
```
#ifdef VK_USE_PLATFORM_WAYLAND_KHR
#include "vulkan_wayland.h"
#endif
```
We can find out the platform only in runtime, so we need to compile the application to work with both platforms.

P.s.
I can't imagine how we can define a windowing system from a CMake.

**Comment by robertosfield** [comment]
I have merged the changes as the branch:

https://github.com/vsg-dev/vsgQt/tree/asafran-WaylandQSurface


Everything compiles fine for me, when I run it it detects xcb so uses that code path which is fine, but I'd like to test the Wayland code path as well.  I have Kubuntu 22.04 installed with RADV RENOIR drivers.  I haven't installed Wayland specific components of Qt yet.

@asafran Do you have a recommendations for setting things up to run with Wayland?

Eventually I'd like the core VSG to support Wayland as well, once that's available I am thinking that may make part of this PR a bit simpler as there shouldn't be any need to change the surface type.  This also applies to Xlib backend.

**Comment by robertosfield** [comment]
I have had a look into how much work it would take to implement Wayland support in the core VSG, one example I came across is reasonably self  contained and only 347 lines of code:

https://github.com/googlesamples/vulkan-basic-samples/blob/master/Sample-Programs/Hologram/ShellWayland.cpp

I would guess this would take 2 days to implement and get working reliably.  I won't be able to dive in and tackle it right away but I'll keep this on my TODO list for some point in the next month or two and if I get a time slot will have bash.  First though I'll need to get Wayland working on this system just with your vsgQt mod.

**Comment by robertosfield** [comment]
@asafran I have refactor ViewerWindow to simplify it so it always relies upon Qt to provide the vkSurface and update how it uses vsg::WindowTraits so that it now works closer to how the core VSG uses recent additions to WindowTraits.

https://github.com/vsg-dev/vsgQt/tree/RefactorInstanceExtensionSurfaceName

The changes include removal of the old VK_USE_PLATFORM_WIN32_KHR etc. definitions, with the instanceExtensionSurfaceName() now being entirely computed on the fly with no #if blocks.

This change does mean that Qt 5.10 or later must be used with is unfortunate but given how much cleaner we can make the code I think it's worth it.  Time has also passed since vsgQt was originally written so supporting these older versions of Qt should be less of an issue.

Could you test out this branch?

**Comment by robertosfield** [comment]
I'm closing this PR, as I feel the [RefactorInstanceExtensionSurfaceName](https://github.com/vsg-dev/vsgQt/tree/RefactorInstanceExtensionSurfaceName) branch will be a cleaner way of tackling this.

**Comment by asafran** [comment]
> I have refactor ViewerWindow to simplify it so it always relies upon Qt to provide the vkSurface and update how it uses vsg::WindowTraits so that it now works closer to how the core VSG uses recent additions to WindowTraits.

I think this is the right decision. I also initially wanted to offer to abandon the possibility of creating a surface on the vsg side.

> Could you test out this branch?

Successfully launched on wayland, window resizing and event handling work fine. It also works fine on X11 session.

> Do you have a recommendations for setting things up to run with Wayland?

Just install qtwayland5, which provides the QPA plugin. In my case, there were no other problems with launching my app on wayland, except those that fix this PR

> I have had a look into how much work it would take to implement Wayland support in the core VSG, one example I came across is reasonably self contained and only 347 lines of code:

I don't know much about wayland, it may be appropriate to use third-party libraries for event handling, because wayland is more complicated in this than X11.

**Comment by robertosfield** [comment]
Will I need to install Wayland based window manager?

I'm using KDE right now which doesn't used Wayland by default.

**Comment by asafran** [comment]
If you are using Kubuntu, you can simply switch the session type in the display manager when logging in. If you are using NVidia, you should enable the DRM kernel mode setting. To enable DRM KMS, add nvidia_drm.modeset=1 as kernel parameter.

https://community.kde.org/KWin/Wayland
https://wiki.archlinux.org/title/NVIDIA#DRM_kernel_mode_setting

**Pull request closed without merging**